Implement Rhombus 2D primitive.

salvadorcarvalhinho opened the pull request:
# Objective

- Create a new 2D primitive, Rhombus, also knows as "Diamond Shape"
- Simplify the creation and handling of isometric projections
- Extend Bevy's arsenal of 2D primitives

## Testing

- New unit tests created in bevy_math/ primitives and bev_math/ bounding
- Tested translations, rotations, wireframe, bounding sphere, aabb and creation parameters

alice-i-cecile commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,119 @@ impl Measured2d for Annulus {
+            if xintercept.abs() * self.half_diagonals.y + yintercept.abs() * self.half_diagonals.x
+                <= diagonals_mult
+            {
+                // The interception betwen the side and the normal is in the perimeter of the rhombus
```
```suggestion
                // The interception between the side and the normal is in the perimeter of the rhombus
```

salvadorcarvalhinho replied:
Thank you, I think we solved this in the recent commit

alice-i-cecile commented:
> Simplify the creation and handling of isometric projections

Can you say more about this? Should we be using this in Bevy's internals?

salvadorcarvalhinho commented:
> > Simplify the creation and handling of isometric projections
> 
> Can you say more about this? Should we be using this in Bevy's internals?

Isometric projection basically is a form of representing 3D objects only with 2D figures and a shape commonly used in this type of projections is the Rhombus. We are proposing to add this because other game engines, like Unity, already have support for this projection and we tought it would be good idea to give this tool to Bevy users. 
Some additional support in Bevy's internals would certainly help the integration of this tool and would enable 3D looking scenes to be rendered with computational power of a 2D scene.

alice-i-cecile reviewed (approved):
Well made and I'm satisfied with the motivation.

IQuick143 commented on crates/bevy_math/src/bounding/bounded2d/primitive_impls.rs:
```diff
@@ -183,6 +183,38 @@ impl Bounded2d for Ellipse {
+    fn aabb_2d(&self, translation: Vec2, rotation: impl Into<Rotation2d>) -> Aabb2d {
+        let rotation_mat = rotation.into();
+
+        let [a, b, c, d] = [
+            rotation_mat * Vec2::new(self.half_diagonals.x, 0.0) + translation,
+            rotation_mat * Vec2::new(0.0, self.half_diagonals.y) + translation,
+            rotation_mat * Vec2::new(-self.half_diagonals.x, 0.0) + translation,
+            rotation_mat * Vec2::new(0.0, -self.half_diagonals.y) + translation,
+        ];
+
+        let min = Vec2::new(
+            a.x.min(b.x).min(c.x).min(d.x),
+            a.y.min(b.y).min(c.y).min(d.y),
+        );
+        let max = Vec2::new(
+            a.x.max(b.x).max(c.x).max(d.x),
+            a.y.max(b.y).max(c.y).max(d.y),
+        );
+
+        Aabb2d { min, max }
```
You can factor out the translations out of the minima and maxima, like so:
```suggestion
        let [a, b, c, d] = [
            rotation_mat * Vec2::new(self.half_diagonals.x, 0.0),
            rotation_mat * Vec2::new(0.0, self.half_diagonals.y),
            rotation_mat * Vec2::new(-self.half_diagonals.x, 0.0),
            rotation_mat * Vec2::new(0.0, -self.half_diagonals.y),
        ];

        let min = Vec2::new(
            a.x.min(b.x).min(c.x).min(d.x),
            a.y.min(b.y).min(c.y).min(d.y),
        ) + translation;
        let max = Vec2::new(
            a.x.max(b.x).max(c.x).max(d.x),
            a.y.max(b.y).max(c.y).max(d.y),
        ) + translation;

        Aabb2d { min, max }
```
Then you can notice that `a = -c` and `b = -d` (as per the linearity of matrix multiplication and the fact that the input vectors are just sign, flipped), meaning we can skip half the matrix multiplications.
```suggestion
        let [a, b] = [
            rotation_mat * Vec2::new(self.half_diagonals.x, 0.0),
            rotation_mat * Vec2::new(0.0, self.half_diagonals.y),
        ];

        let min = Vec2::new(
            a.x.min(b.x).min(-a.x).min(-b.x),
            a.y.min(b.y).min(-a.y).min(-b.y),
        ) + translation;
        let max = Vec2::new(
            a.x.max(b.x).max(-a.x).max(-b.x),
            a.y.max(b.y).max(-a.y).max(-b.y),
        ) + translation;

        Aabb2d { min, max }
```
Taking the maximum between `a` and `-a` amounts to taking the absolute value.
Taking the minimum between `a` and `-a` amounts to taking minus the absolute value.
```suggestion
        let [a, b] = [
            rotation_mat * Vec2::new(self.half_diagonals.x, 0.0),
            rotation_mat * Vec2::new(0.0, self.half_diagonals.y),
        ];

        let min = Vec2::new(
            (-a.x.abs()).min(-b.x.abs()),
            (-a.y.abs()).min(-b.y.abs()),
        ) + translation;
        let max = Vec2::new(
            (a.x.abs()).max(b.x.abs()),
            (a.y.abs()).max(b.y.abs()),
        ) + translation;

        Aabb2d { min, max }
```
Factoring out the minus signs by realising `min(-x, -y) = -max(x,y)`, we find out that min and max are basically the same up to a sign.
We can also simplify the operations by using the elementwise operations on Vec2's.
```suggestion
        let [rotated_x_half_diagonal, rotated_y_half_diagonal] = [
            rotation_mat * Vec2::new(self.half_diagonals.x, 0.0),
            rotation_mat * Vec2::new(0.0, self.half_diagonals.y),
        ];

        let aabb_half_extent = rotated_x_half_diagonal.abs().max(rotated_y_half_diagonal.abs());

        Aabb2d { -aabb_half_extent + translation , aabb_half_extent + translation }
```
IIRC there was an even cleverer trick, that simply took the absolute value of the matrix [I think it's used in rectangle code].
But I'm not so sure about its derivation right now.

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,119 @@ impl Measured2d for Annulus {
+    /// Get the length of each side of the rhombus
+    #[inline(always)]
+    pub fn side(&self) -> f32 {
+        (self.half_diagonals.x.powi(2) + self.half_diagonals.y.powi(2)).sqrt()
```
There's the `f32::hypot` function, idk if it's necessarily worth using.

Jondolf replied:
Isn't this equivalent to just `self.half_diagonals.length()`?

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,119 @@ impl Measured2d for Annulus {
+    #[inline(always)]
+    #[doc(alias = "apothem")]
+    pub fn inradius(&self) -> f32 {
+        let horizontal_diagonal = self.half_diagonals.x * 2.0;
+        let vertical_diagonal = self.half_diagonals.y * 2.0;
+        (horizontal_diagonal * vertical_diagonal)
+            / (2.0 * (horizontal_diagonal.powi(2) + vertical_diagonal.powi(2)).sqrt())
```
```suggestion
        let horizontal_diagonal = self.half_diagonals.x * 2.0;
        let vertical_diagonal = self.half_diagonals.y * 2.0;
        (horizontal_diagonal * vertical_diagonal)
            / (2.0 * (horizontal_diagonal.powi(2) + vertical_diagonal.powi(2)).sqrt())
```
Expanding out the definition for `horizontal|vertical_diagonal`, we arrive at:
```suggestion
        (self.half_diagonals.x * self.half_diagonals.y) / self.side()
```
Geometrical interpretation is that `self.half_diagonals.x * self.half_diagonals.y` is `2*Area` of one quarter right triangle of the rhombus. The inradius of the rhombus is just the height of this triangle opposite to the hypotenuse, hence it's `2*Area / Hypotenuse`.

Also note:
This panics when the rhombus is 0, but the method should probably just return 0 in that case.

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,119 @@ impl Measured2d for Annulus {
+    /// Get the area of the rhombus
+    #[inline(always)]
+    fn area(&self) -> f32 {
+        (self.half_diagonals.x * 2.0) * (self.half_diagonals.y * 2.0) / 2.0
```
Maybe we can just reduce this to `2.0 * self.half_diagonals.x * self.half_diagonals.y`?

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,119 @@ impl Measured2d for Annulus {
+    /// Get the perimeter of the rhombus
+    #[inline(always)]
+    fn perimeter(&self) -> f32 {
+        2.0 * ((self.half_diagonals.x * 2.0).powi(2) + (self.half_diagonals.y * 2.0).powi(2)).sqrt()
```
And this could be just `4.0 * self.side()`.

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -1618,6 +1745,14 @@ mod tests {
+        assert_eq!(rhombus.area(), 6.0, "incorrect area");
+        assert_eq!(rhombus.perimeter(), 10.0, "incorrect perimeter");
+        assert_eq!(rhombus.side(), 2.5, "incorrect side");
+    }
```
There should be a test for the `inradius`.

salvadorcarvalhinho commented:
@IQuick143 Thanks a lot for the feedback, it really helped to simplify the code! We believe we already implemented all of the suggested changes in our latest commit. 
We also added new tests to verify that everything is working correctly and added new constructors to the Rhombus (similarly to what is done in the rectangle).

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,133 @@ impl Measured2d for Annulus {
+    #[inline(always)]
+    pub fn from_side(side: f32) -> Self {
+        Self {
+            half_diagonals: Vec2::new(side.hypot(side) / 2.0, side.hypot(side) / 2.0),
```
```suggestion
            half_diagonals: Vec2::splat(side * std::f32::consts::FRAC_1SQRT_2),
```

salvadorcarvalhinho replied:
When we tried this solution the results were just some decimal places changed (for example, instead of 1.0 it was 0.999994), so, in our last commit, we decided to use the `splat` but with the `hypot` so the precision is as high as possible.

IQuick143 replied:
That is odd... I guess it's not a huge deal in a constructor. Btw `0.999994` is pretty much within f32 precision (about 7 decimal places) IIRC.

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,133 @@ impl Measured2d for Annulus {
+    #[inline(always)]
+    #[doc(alias = "apothem")]
+    pub fn inradius(&self) -> f32 {
+        ((self.half_diagonals.x * self.half_diagonals.y) / self.side()).max(0.0)
```
This resolves the issue with the division by 0? Didn't know that's possible.

Jondolf replied:
`x / 0.0` is positive infinity in Rust iirc, so I'm pretty sure this does *not* work

salvadorcarvalhinho replied:
When testing, the result of the division by 0 appeared as `NaN` to us, and when we used this method it was working correctly. But it really is a little bit weird so we changed the function and added the `if` statement, in our latest commit.

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,133 @@ impl Measured2d for Annulus {
+    #[inline(always)]
+    pub fn closest_point(&self, point: Vec2) -> Vec2 {
+        let diagonals_mult = self.half_diagonals.x * self.half_diagonals.y;
+
+        // Point is inside the rhombus
+        if point.x.abs() * self.half_diagonals.y + point.y.abs() * self.half_diagonals.x
+            <= diagonals_mult
+        {
+            point
+        } else {
+            // Equation of the line that contains the side in the points quadrant
+            let [m, b] = [
+                (-point.x * point.y).signum() * self.half_diagonals.y / self.half_diagonals.x,
+                point.y.signum() * self.half_diagonals.y,
+            ];
+
+            let inv_m = -1.0 / m;
+
+            // Normal of the line mx+b that passes through the point
+            let side_normal = [inv_m, point.y - inv_m * point.x];
+
+            // Intersection of the side normal and the side line rounded to 4 decimal places
+            let xintercept =
+                (((b - side_normal[1]) / (side_normal[0] - m)) * 10000.0).round() / 10000.0;
+            let yintercept = ((m * xintercept + b) * 10000.0).round() / 10000.0;
+
+            if xintercept.abs() * self.half_diagonals.y + yintercept.abs() * self.half_diagonals.x
+                <= diagonals_mult
+            {
+                // The interception between the side and the normal is in the perimeter of the rhombus
+                Vec2::new(xintercept, yintercept)
+            } else {
+                // Point is in one of the axys or close to it
+                if point.x.abs() > point.y.abs() {
+                    Vec2::new(point.x.signum() * self.half_diagonals.x, 0.0)
+                } else {
+                    Vec2::new(0.0, point.y.signum() * self.half_diagonals.y)
+                }
+            }
+        }
```
This code is untested, so please check if I wrote it correctly.
But I think it's a cleaner and better way to handle this.
```suggestion
// Fold the problem into the positive quadrant
let point_abs = point.abs();
let half_diagonals = self.half_diagonals.abs(); // to ensure correct sign
// The unnormalised normal vector perpendicular to the side of the rhombus
let normal = Vec::new(half_diagonals.y, half_diagonals.x)
let normal_magnitude_squared = normal.length_squared();
if normal_magnitude_squared == 0.0 { // consider making this comparison a bit more lenient
    return Vec2::ZERO; // A null Rhombus has only one point anyway.
}
// The last term corresponds to normal.dot(rhombus_vertex)
let distance_unnormalised = normal.dot(point_abs) - half_diagonals.x * half_diagonals.y;
// The point is already inside so we simply return it.
if distance_unnormalised <= 0 {
    return point;
}

// Clamp the point to the edge
let mut result = point_abs - normal * distance_unnormalised / normal_magnitude_squared;
// Clamp the point back to the positive quadrant
// if it's outside, it needs to be clamped to either vertex
if result.x <= 0 {
    final_point = Vec2::new(0.0, half_diagonals.y);
}
if result.y <= 0 {
    final_point = Vec2::new(half_diagonals.x, 0.0);
}
// Finally, we restore the signs of the original vector
result.copysign(point)
```

salvadorcarvalhinho replied:
The code was perfect, and we only needed to change some little things. This is really clever and works marvelously, thanks a lot for the help!

IQuick143 commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -1601,6 +1728,20 @@ mod tests {
 
+    #[test]
+    fn rhombus_closest_point() {
+        let rhombus = Rhombus::new(2.0, 1.0);
```
For completeness, we should add a test for a null rhombus (0.0, 0.0)

salvadorcarvalhinho replied:
Yes, that wasn't tested, so we added the test in our latest commit

Jondolf commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,133 @@ impl Measured2d for Annulus {
+    /// Get the length of each side of the rhombus
+    #[inline(always)]
+    pub fn side(&self) -> f32 {
+        self.half_diagonals.x.hypot(self.half_diagonals.y)
```
This is currently the squared length, right? This would need a `sqrt` like before, but it's also just equivalent to using `length` for this
```suggestion
        self.half_diagonals.length()
```

Jondolf replied:
oh sorry no, this is correct too, I forgot how `hypot` works 😅 but `length` might still be faster

salvadorcarvalhinho replied:
You are right, the `lenght` function works and is much simpler. Thanks for the suggestion!

alice-i-cecile commented:
@salvadorcarvalhinho feel free to resolve conversations as you address them :) It makes it easier for reviewers to track what work remains to be done.

Jondolf commented on crates/bevy_math/src/primitives/dim2.rs:
```diff
@@ -929,6 +929,132 @@ impl Measured2d for Annulus {
+    #[inline(always)]
+    #[doc(alias = "apothem")]
+    pub fn inradius(&self) -> f32 {
+        let side = self.side();
+        if side == 0.0 {
+            0.0
+        } else {
+            (self.half_diagonals.x * self.half_diagonals.y) / self.side()
+        }
```
Could avoid computing `.side()` twice here
```suggestion
        let side = self.side();
        if side == 0.0 {
            0.0
        } else {
            (self.half_diagonals.x * self.half_diagonals.y) / side
        }
```

salvadorcarvalhinho replied:
Oh, I forgot to change it bellow, the declaration was made already thinking in that 🤦🏽‍♂️
Thank you!

IQuick143 reviewed (approved):
Unless I overlooked something, I think the code's good now.
